mint-client/src/clients/gateway.rs

```diff
@@ -226,6 +303,40 @@ impl GatewayClient {
             .collect()
     }
 
+    pub async fn await_preimage_decryption(&self, txid: TransactionId) -> Result<Preimage> {
+        loop {
+            match self.context.api.fetch_tx_outcome(txid).await {
+                Ok(status) => match status {
+                    TransactionStatus::Accepted { outputs, .. } => {
+                        match &outputs[0] {
+                            OutputOutcome::LN(minimint::modules::ln::OutputOutcome::Contract {
+                                outcome,
+                                ..
+                            }) => {
+                                match outcome {
+                                    ContractOutcome::Incoming(DecryptedPreimage::Some(
+                                        preimage,
+                                    )) => return Ok(preimage.clone()),
+                                    ContractOutcome::Incoming(DecryptedPreimage::Pending) => {}
+                                    ContractOutcome::Incoming(DecryptedPreimage::Invalid) => {
+                                        return Err(GatewayClientError::InvalidPreimage)
+                                    }
+                                    _ => return Err(GatewayClientError::WrongContractType),
+                                };
+                            }
+                            _ => return Err(GatewayClientError::WrongTransactionType),
+                        };
+                    }
+                    TransactionStatus::Error(error) => {
+                        return Err(GatewayClientError::InvalidTransaction(error))
+                    }
+                },
+                Err(error) => return Err(GatewayClientError::MintApiError(error)),
+            };
+            tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
+        }
+    }
```

Comment:
I implemented this way to preserve the different types of errors. But this code is really ugly. Tempted to just match on `Ok(TransactionStatus::Accepted(OutputOutcome::Ln( ... )))` and treat other errors as all the same ...